src/file/file.rs

```diff
@@ -0,0 +1,282 @@
+use std::{fs, path::PathBuf};
+
+use super::{ArgumentToken, Expression, File, Line};
+
+pub fn read_ssh_config(path: &PathBuf) -> std::io::Result<File> {
+    fs::read_to_string(path).map(|content| parse_ssh_config(content.as_str(), Some(path.clone())))
+}
+
+pub fn parse_ssh_config(content: &str, path: Option<PathBuf>) -> File {
+    let mut lines = Vec::<Line>::new();
+    for line in content.lines() {
+        lines.push(parse_line(line));
+    }
+    File { lines, path }
+}
+
+pub fn write_ssh_config(file: &File, path: PathBuf) -> std::io::Result<()> {
+    fs::write(path, file.to_string())
+}
+
+fn parse_line(line: &str) -> Line {
+    let content = line.trim_end_matches(['\n', '\r']);
+    let mut it = content.chars();
+    if it.clone().any(|c| c == '\n') {
+        panic!("multiline string can not be parsed as a single line");
+    }
+    let indent_prefix: String = it.by_ref().take_while(|s| s.is_whitespace()).collect();
+    let indent_suffix_len = it.by_ref().rev().take_while(|s| s.is_whitespace()).count();
+    let indent_suffix: String = content[(content.len() - indent_suffix_len)..].to_string();
+
+    Line {
+        indent_prefix,
+        expression: parse_expression(line.trim()),
+        indent_suffix,
+    }
+}
+
+fn parse_expression(content: &str) -> Expression {
+    if content.is_empty() {
+        return Expression::Empty;
+    } else if content.starts_with('#') {
+        return Expression::Comment(content.to_string());
+    }
+
+    let keyword: String = content.chars().take_while(|c| c.is_alphabetic()).collect();
+
+    if keyword.is_empty() {
+        return Expression::Malformed(content.to_string());
+    }
+
+    let separator: String = content
+        .chars()
+        .skip(keyword.len())
+        .take_while(|c| c.is_whitespace() || c == &'=')
+        .collect();
+
+    if !separator.is_empty()
+        && (separator.as_str().trim().is_empty() || separator.as_str().trim() == "=")
+    {
+        if let Some(arguments_expression) =
+            parse_arguments_expression(content[(keyword.len() + separator.len())..].trim())
+        {
+            return Expression::ConfigurationOptions {
+                keyword,
+                separator,
+                arguments_expression,
+            };
+        } else {
+            return Expression::Malformed(content.to_string());
+        }
+    }
+
+    Expression::Malformed(content.to_string())
+}
+
+fn parse_arguments_expression(content: &str) -> Option<Vec<ArgumentToken>> {
+    let mut arguments_expression = Vec::<ArgumentToken>::new();
+    let mut remaining = content;
+
+    while !remaining.is_empty() {
+        if remaining.starts_with(|c: char| c.is_whitespace()) {
+            let argument: String = remaining
+                .chars()
+                .take_while(|c| c.is_whitespace())
+                .collect();
+            remaining = &remaining[argument.len()..];
+            arguments_expression.push(ArgumentToken::Whitespace(argument));
```

Comment:
I suggest to do just collection whitespaces on each iteration and if there is an at least 1 whitespace collected then emit whitespace token.